src/ripple/app/misc/NegativeUNLVote.h

```diff
@@ -0,0 +1,193 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+#define RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/protocol/Protocol.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/UintTypes.h>
+
+#include <optional>
+
+namespace ripple {
+
+class Ledger;
+template <class Adaptor>
+class Validations;
+class RCLValidationsAdaptor;
+using RCLValidations = Validations<RCLValidationsAdaptor>;
+class SHAMap;
+namespace test {
+class NegativeUNLVoteInternal_test;
+class NegativeUNLVoteScoreTable_test;
+}  // namespace test
+
+/**
+ * Manager to create NegativeUNL votes.
+ */
+class NegativeUNLVote final
+{
+public:
+    /**
+     * A validator is considered unreliable if its validations is less than
+     * nUnlLowWaterMark in the last flag ledger period.
+     * An unreliable validator is a candidate to be disabled by the NegativeUNL
+     * protocol.
+     */
+    static constexpr size_t nUnlLowWaterMark = 256 * 0.5;
```

Comment:
I agree that sticking to integer arithmetic works; we can always write `INT * FLOAT` as `INT * INT / INT` for almost every `FLOAT` value we care about here:

| Floating Expression | Equivalent Integer |
|:------------------:|:-------------------------:|
|`256 * 0.5` | `256 * / 2` |
|`256 * 0.8` | `(256 * 4) / 5` |
|`256 * 0.9` | `(256 * 9) / 10` |

I don't _think_ the C++ standard actually mandates IEEE floating-point math⁠ but I don't think that generally "bites" us for the numbers we're using—perhaps @HowardHinnant or @scottschurr can correct me. If it does, then we should use the floating point arithmetic format, since it's "easier" to read: I understand *`0.8`* as "80% percent" without having to think. I don't have the same "intuitive" response to *`4 / 5`*. 

As a sidenote, and only because of the specific numbers involved, we should consider moving away from this duocentehexaquinquagesimal numbers… there's not much point in it, and using decimal here makes a lot more sense, especially since we're talking percentages and those are best expressed in that base.